### ISSUE TITLE
Any new release / pypi release scheduled for Beatbox?

### ISSUE THREAD

**Issue by OakNinja** [superfell/Beatbox#43]
Is there any release planned for Beatbox? Mainly I miss python 3 support in the current pypi release. Is there anything I can do to help?

**Comment by hynekcer** [comment]
Yes, someone who knows English well can help. I wrote support Python 3 to two repositories This one original repository and [davisagli/beatbox](https://github.com/davisagli/beatbox) that is the version on pypi. davisagli Davis Glick is a very active developer but he is very busy and did not responded me to a very old issue, maybe it was not well English (the only issue in that repository) The Python 3 support works also (here https://github.com/hynekcer/beatbox-davisagli) I dont care which version will continue on PyPI. I have many new unpublished features for the current Salesforce but without PyPI it's not right and the feedback is lost.

**Comment by OakNinja** [comment]
Thanks @hynekcer. I guess this repo is the most up to date? I'll gladly help with whatever is needed to get a new release out. Would be great to hear what @davisagli and @superfell think as well.

**Comment by hynekcer** [comment]
@OakNinja It is very improbable that davisagli can read a "@" from this repository and also understand the context at all. He doesn't watch this repository, didn't star, did not clone and maybe has never visited this. You should write something short, clear, specific, addressed directly to him, so smart that he can do something simple and useful for beatbox.

> is this repo is the most up to date?

This repository is the most active and with the best chance to get attention and contributions. It is not possible to say absolutely "the most up to date" about any one. The forks of beatbox don't share usually any new code in no direction, don't get any feedback and they looks sometimes expecting to have only a personal value, not a general. A very important clone group by alanjcastonguay (the only group) was motivated also by the problem with PyPI. They implemented notable features. They would prefer to participate on any solution if the problem would be unblocked.

My goal is that the original author Superfell would be added to maintainers of Beatbox and Davisagli will experience enough thanks to him. (what I found about [package Owners and Maintainers on PyPI] (https://docs.python.org/2/distutils/packageindex.html#pypi-overview))
A good probe of responsiveness is to send a pull request. I'm going to do it

**Comment by alanjcastonguay** [comment]
>A very important clone group by alanjcastonguay (the only group) was motivated also by the problem with PyPI. They implemented notable features. They would prefer to participate on any solution if the problem would be unblocked.

Reporting in, Captain. https://github.com/alanjcastonguay/pyforce would ideally be slimmed down to a object-marshaling wrapper with a dependency on beatbox for all transit. My day job isn't demanding this, so it became depriorized.

**Comment by davisagli** [comment]
Hello all and sorry for the radio silence. I've been using the Salesforce REST API lately (via https://pypi.python.org/pypi/simple-salesforce) so am not really interested in maintaining the fork of beatbox that was developed at my former employer, Groundwire (which is the one published at https://pypi.python.org/pypi/beatbox).

I'm willing to hand over control of the "beatbox" name on PyPI if there's consensus about who it should go to. In some ways it sounds like @alanjcastonguay would be a natural choice, since his fork has carried on in the same direction with the Python wrapper that was started by Groundwire. On the other hand, it sounds like @hynekcer is the only one of us actively motivated to work on development. On the third hand, I feel bad that we forked @superfell's original work without renaming it (sorry Simon, I didn't realize at the time what a faux pas that is), so I'd like to give him the right of first refusal.

So what should I do with the PyPI ownership?

(Regardless of who gets it, I'd like to request that none of the existence released packages be deleted (to avoid breaking existing deployments) and that a message be posted explaining the situation.)

**Comment by superfell** [comment]
Thanks everyone for their comments, trying to summarize where we are
sf/beatbox -> forked to dg/beatbox -> forked to ac/pyforce
all three forks have post fork changes

in addition, dg/beatbox is on pypi at pypi/beatbox, and ac/pyforce is at pypi/pyforce

I've never used pypi so am unclear on the expectations of its users or package authors, but presumably a newer version of pypi/beatbox should be roughly compatible with the current version i.e. we shouldn't publish sf/beatbox in its current form as a new version at pypi/beatbox

Options [in no particular order].
1. Do nothing. we've all lived with this like this for a while, accept the fact the SOAP API usage is declining, this issue will naturally die out eventually. 

2. collapse the folks to ac/pyforce, mostly likely by updating pypi/beatbox telling people to use pypi/pyforce going forward and taking the current state of sf/beatbox and merging it into ac/pyforce. All future work is based on that version.

3. similar to (2), collapse the forks back to sf/beatbox, publish the resulting new version at pypi/beatbox

4. collapse dg/beatbox to ac/pyforce, update pypi/beatbox to tell people to switch to pyforce, live with the fact that its divergent from sf/beatbox. 

5. collapse ac/pyforce to dg/beatbox, give control of pypi/beatbox to AC, live with the pypi/beatbox != sf/beatbox issue [or do 6 as well].

6. rename sf/Beatbox so it can be published to PyPi, have ac/pyforce and/or dg/beatbox take that as a dependency rather than a fork. While this bugs me as the original author, i think most people think of beatbox as the pypi version, so I'd live with it.

Options 2 or 3 would seem like the best outcome for users of the library, I don't have a good sense for how much work is involved, and like everyone else i don't have a lot of time available to work on it.

For the PyPi folks, are there usage stats, any idea on whether pypi/beatbox or pypi/pyforce is the most used ? [this would be useful input for 2 vs 3].

Not sure options 4 or 5 make sense unless 6 is done as well.

Other options, strong opinion for or against? @hynekcer you've been the most active maintainer recently, do you have a preference for the outcome ?

**Comment by alanjcastonguay** [comment]
PyPi download stats (as of 2016-05-31?)

* 92,313 http://pypi-ranking.info/module/beatbox
* 36,729 http://pypi-ranking.info/module/Products.salesforcebaseconnector 
* 25,131 http://pypi-ranking.info/module/pyforce
* 22,917 http://pypi-ranking.info/module/beatboxxx
* 1,406 http://pypi-ranking.info/module/beatbox3

**Comment by hynekcer** [comment]
S.F., thanks for a comprehensive answer and thanks also you @davisagli and @alanjcastonguay.
It is great that we can cooperate some way and willing to agree on searching a solution even if nobody of us has much time. I can not presume a realistic final state that can be achieved soon. A wise and relative easy first step should be done now, without grandiose promises.

I understand that users are expecting currently a new version especially with Python 3. Every from these 3 projects can inform about all them in the README that some form of fusion or binding is probable.

I prefer to release the version 37.0 from the D.G. repository and to inform that is probably the last major version from this repository and only bugfixes will be done for one year, mostly by me if nobody contribute. Eventual new feature requests should be directed the S.F. repository where I can work on a new version expected next year, very compatible with the current D.G. version. S.F. wrote an important note, that tests should be added to the code committed by me at July. This is a reason why it was easier to do a small change to D.G. code. I can promise that the current tests by D.G would pass also on the new version next year and that no additional dependencies would be required for run time. (though some optional dependencies are useful and that some dependencies will be necessary for tests.) For pyforce I can imagine an interchange of code with A.C., to the extent that it will not block a new development.

I have also no experience with PyPI package administration, but it should not be a problem. I think that the package at PyPI should have two owners now, including D.G.

I will write tomorrow more about my intentions and dubieties. (similarly as S.F wrote about one)

A.C: I knew that PyPI *beatbox* had been the most popular package from the related family and that its "secret" source repository at DG had the lowest possible popularity [1, 0, 1] when only the author knows about it, while sf/beatbox repository is the most popular in the family.

**Comment by hynekcer** [comment]
First something about me, my intentions and doubt.
My goal is to use Beatbox as a *solid base* for an OSS project that I would like to start.
Dubiety: I hope that some form of grant can be found, but I can imagine only from Salesforce, because they can probably profit from it enough. It is possible that I will continue to develop Beatbox hard in the original focus, eventually I would have very little time or I don't find funding and lose a motivation.

My idea is:
1) Create a Python DB API compatible client for Salesforce that supports also classic SQL instead of only API and SOQL: SELECT FROM JOIN etc., INSERT INTO, UPDATE tbl SET, DELETE FROM. The benfit is to simplify the process of writing SOQL, to create a base for connection to different ORM tools and to create a more efficient SOQL that is usually written by hand. The most complicated part is the compiler of JOIN based queries to SOQL. The input would be use `describeSObject`, alternatively the exported `Enterprise WSDL` xml. It is sometimes possible to compile more SQL queries to one SOQL and to parse the result to more cursors back, because SOQL is very powerful for some types of tasks.

2) Write a client with the same functionality in another language (to inspire other people in more languages), that use a short runtime in that language and also "static" queries SOQL precompiled at developers machine by Python or precompiled by an equivalent web service.

I would like to estimate a possible interest of application developers, especially if it can attract attention inside Salesforce. (complicated - no idea currently)

I wrote important steps to prove the concept in the project Django-Salesforce and now I feel a turning point to unfold it for general use and get a fund for it or to stop development. (If I would continue on a small playgroung python-django to in that portable direction, with so friendly license and don't open it enough to be interesting for Salesforce, someone other can start a great game.)


Consequences for Beatbox - if it will continue:
- General authentication and general connection factory. It is a small change and one additional parameter in the initialization of Client, but it allows to create powerful extensions that can be common for more Force.com APIs (e.g. REST and BULK), not only for SOAP. The same authentication token and the same network connection can be shared between them.
- There is a small issue with Python 3 that a warning is reported if the connection is repeatedly closed and created. I don't think that it is worse than in Python 2, it was only not reported. A network factory with `requests` package uses a connection pool and it is better for multi-tenant applications. It is not necessary to be present in Beatbox, only to be modular.
- I have code for "mock" testing that the network communication of a test can be recorded and replayed fast the responses next time at the next run if the test requests did not changed. It is useful if a code at higher level above Beatbox should be tested fast. It can be useful also for unit tests in Beatbox. The test mode can be switched between record, playback and combined mode, where the playback is compared with a new live communication.


S.F. You are right about the distant prospect of SOAP. The  REST API is improved every year and it is still capable to transfer less amount of data, shorter queries etc. They only fill the most painful gaps, e.g. `composite/tree` REST requests since API 34.0 for inserting up to 200 rows by one request, but nothing for middle range update or delete. The REST is still sometimes too lightweight and BULK is too heavy weight.

I have reviewed the pyforce code in August. If I remember a little, there were more some modifications that looked problematic, I have a very good marshaling in Django-salesforce and it was not crucial, also not sending emails etc. I also felt that no consolidation is possible without start with S.F.

**Comment by OakNinja** [comment]
To me, it seems that @hynekcer would be the best maintainer given the above information, if he's up for it. What do @superfell @alanjcastonguay  @davisagli think?

**Comment by alanjcastonguay** [comment]
I am unclear whether @hynekcer is proposing breaking changes to the libraries' published interfaces (discouraging), or maintaining the existing interfaces and building stuff on top (cool).

**Comment by hynekcer** [comment]
I can promise compatibility with the latest beatbox 32.1 and also no headless jump into pyforce, besides I have no way how to test beatbox with pyforce myself.

**Comment by alanjcastonguay** [comment]
Sounds good to me.

+1 - I am happy to make @hynekcer maintainer of all repos and pypi releases.

There's no way to use, let alone test, beatbox with pyforce today. There's a vendored copy of ancient beatbox embedded inside. Though it is likely possible to rip that out and import modern beatbox instead. Hopefully with a minimum of change. Then it becomes a matter of testing pyforce's marshalling while mocking out beatbox, or testing them together.

**Comment by alanjcastonguay** [comment]
@hynekcer do you want to own the pyforce package?

**Comment by hynekcer** [comment]
No, thanks. It is more that I can ever use. I couldn't be a maintainer of pyforce, because I want continue to use Beatbox - SOAP API in Django-salesforce package. I did like to fix Beatbox and I was grateful you for the constructive communication.

I use or want to use more advanced authentication methods, introspection, error handling, test response injection etc. compatible with Django-salesforce and I use both even on the same network connection. I need therefore a very low level access.
It is even good accepted that the used Beatbox dependency is realized by a long Github link instead of short PyPI name. My interest can be later revamped if a significant opportunity arises, but it seems realistic in Beatbox range.

**Comment by hynekcer** [comment]
I think that everything important can be now solved also by REST API without Beatbox. The most important advance in Salesforce REST API (to be almost complete) is by [Composite Resources](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_composite.htm).

> **Composite** Executes a series of REST API requests in a single call. You can use the output of one request as the input to a subsequent request. The response bodies and HTTP statuses of the requests are returned in a single response body. The entire request counts as a single call toward your API limits.

or an introduction to it by [blog post about Composite](https://developer.salesforce.com/blogs/tech-pubs/2017/01/simplify-your-api-code-with-new-composite-resources.html). The simple new `composite/` resource is better than the older composite/batch/ or composite/tree/.

Other important solution that I didn't know previously is how to [Convert Lead by REST API](https://stackoverflow.com/questions/19663119/is-there-any-rest-service-available-in-salesforce-to-convert-leads-into-accounts)

(That is why I will not use and contribute to Beatbox more. Thanks this great project.)

**Comment by OakNinja** [comment]
Closing this as the REST API is more convenient.
